### ISSUE TITLE
Timeout and Errors, devices no more working after Internet drop out

### ISSUE THREAD

**Issue by StephaneM60** [albertogeniola/meross-homeassistant#263]
Hi,
I'm in the process of switching from Domoticz to Home Assistant and I'm trying to use your implementation of Meross devices. Previously on Domoticz I was using a similar approach (Node JS + MQTT) with a much less complex API and it has always worked flawlessly (I have 14 meross devices : light bulb, garage opener, socket...)

This library is awsome when it works but seems to be very sensitive to network stability.

For instance : on sunday night I restart my router, so for a few minutes, no more Internet and no more Wifi, so it's fair to think that all meross devices during this black out cannot work, but I expect them to work again once everything is up an running again.

With the Domoticz plugin it has never been a problem (in fact when the router restarts, I can't see the plugin complaining, it does encounter some timout issue, and I can also see in its log some "Appliance.System.Online" status when device connects / disconnect from : I do have devices that switch from one access point to another)

With your Home Assistant Plugin : following the router restart, all devices failed to work again (they are seen as available, but no more energy data coming and turning on/off etc... devices is not working eitther so communication is broken in both directions)

**Your environment**
HomeAssistant version: Home Asisstant OS 6.2 on Hyper V VM / core-2021.8.8
Hassio Version (if applicable): supervisor-2021.06.8
Hardware environment: VM 2 cores / 2 GB RAM

**Logs taken when the issue happened**
```
Log from Energy device when router is rebooting

2021-08-29 04:45:28 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.prise_salle_a_manger_mss310_power_sensor_w_0 is taking over 10 seconds
2021-08-29 04:45:28 ERROR (MainThread) [meross_iot.manager] Timeout occurred while waiting a response for message b'{"header": {"from": "/app/55147-dce9f6b40e9066af8782277ad98dd10d/subscribe", "messageId": "918b7e8875d753337bc55050210e8832", "method": "GET", "namespace": "Appliance.Control.Electricity", "payloadVersion": 1, "sign": "ff44e84231420bd26789b018edafcaa1", "timestamp": 1630205119}, "payload": {"channel": 0}}' sent to device uuid 18122136570906251h0134298f1450de. Timeout was: 10.0 seconds. Mqtt Host: mqtt-eu.meross.com:443.Global manager stats (last minute): Issued -> 16, Delayed -> 0, Dropped -> 0
2021-08-29 04:45:28 ERROR (MainThread) [custom_components.meross_cloud.sensor] Error occurred.
-------------------------------------
Component version: 0.4.2.2
Device info: 
	Name: Prise Salle à Manger
	UUID: 18122136570906251h0134298f1450de
	Type: mss310
	HW Version: 2.0.0
	FW Version: 2.1.12
Error Message: "An exception occurred"
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 919, in _async_send_and_wait_ack
    return await asyncio.wait_for(future, timeout, loop=self._loop)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/meross_cloud/sensor.py", line 330, in async_update
    await self._device.async_get_instant_metrics(channel=self._channel_id)
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/mixins/electricity.py", line 43, in async_get_instant_metrics
    result = await self._execute_command(method="GET",
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/device.py", line 281, in _execute_command
    return await self._manager.async_execute_cmd(destination_device_uuid=self.uuid,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 838, in async_execute_cmd
    return await self.async_execute_cmd_client(client=client,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 903, in async_execute_cmd_client
    response = await self._async_send_and_wait_ack(
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 934, in _async_send_and_wait_ack
    raise CommandTimeoutError(message=str(message), target_device_uuid=target_device_uuid, timeout=timeout)
meross_iot.model.exception.CommandTimeoutError
```
Log when router is up and running 
```
2021-08-29 07:06:54 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.routeur_internet_mss310_power_sensor_a_0 is taking over 10 seconds
2021-08-29 07:06:54 ERROR (MainThread) [meross_iot.manager] Timeout occurred while waiting a response for message b'{"header": {"from": "/app/55147-dce9f6b40e9066af8782277ad98dd10d/subscribe", "messageId": "d76a492c8f8bb619a39a27bf010bf107", "method": "GET", "namespace": "Appliance.Control.Electricity", "payloadVersion": 1, "sign": "e5c918b6ed28fe77ac7c19e161052104", "timestamp": 1630213605}, "payload": {"channel": 0}}' sent to device uuid 18071993274327251a0534298f1371ee. Timeout was: 10.0 seconds. Mqtt Host: mqtt-eu.meross.com:443.Global manager stats (last minute): Issued -> 5, Delayed -> 0, Dropped -> 0
2021-08-29 07:06:54 ERROR (MainThread) [custom_components.meross_cloud.sensor] Error occurred.
-------------------------------------
Component version: 0.4.2.2
Device info: 
	Name: Routeur Internet
	UUID: 18071993274327251a0534298f1371ee
	Type: mss310
	HW Version: 1.0.0
	FW Version: 1.1.18
Error Message: "An exception occurred"
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 919, in _async_send_and_wait_ack
    return await asyncio.wait_for(future, timeout, loop=self._loop)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/meross_cloud/sensor.py", line 365, in async_update
    await self._device.async_get_instant_metrics(channel=self._channel_id)
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/mixins/electricity.py", line 43, in async_get_instant_metrics
    result = await self._execute_command(method="GET",
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/device.py", line 281, in _execute_command
    return await self._manager.async_execute_cmd(destination_device_uuid=self.uuid,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 838, in async_execute_cmd
    return await self.async_execute_cmd_client(client=client,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 903, in async_execute_cmd_client
    response = await self._async_send_and_wait_ack(
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 934, in _async_send_and_wait_ack
    raise CommandTimeoutError(message=str(message), target_device_uuid=target_device_uuid, timeout=timeout)
meross_iot.model.exception.CommandTimeoutError
```
Log when trying to turn on a switch
```
2021-08-29 09:11:55 ERROR (MainThread) [meross_iot.manager] Timeout occurred while waiting a response for message b'{"header": {"from": "/app/55147-dce9f6b40e9066af8782277ad98dd10d/subscribe", "messageId": "23221528ed1c01d0bafd6976c1e156f3", "method": "SET", "namespace": "Appliance.Control.ToggleX", "payloadVersion": 1, "sign": "47db47bcbefd9ae1070ef5e9237bca52", "timestamp": 1630221105}, "payload": {"togglex": {"onoff": 1, "channel": 0}}}' sent to device uuid 19082032914180258h0848e1e900924a. Timeout was: 10.0 seconds. Mqtt Host: mqtt-eu.meross.com:443.Global manager stats (last minute): Issued -> 7, Delayed -> 0, Dropped -> 0
2021-08-29 09:11:55 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall switch.turn_on (c:370c6a53690fd1f8f0d0cec558878fcf): entity_id=['switch.applique_office_mss710_main_channel']>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 919, in _async_send_and_wait_ack
    return await asyncio.wait_for(future, timeout, loop=self._loop)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/core.py", line 1507, in catch_exceptions
    await coro_or_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1526, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 658, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 856, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 695, in _handle_entity_call
    await result
  File "/config/custom_components/meross_cloud/switch.py", line 103, in async_turn_on
    await self._device.async_turn_on(channel=self._channel_id, skip_rate_limits=True)
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/mixins/toggle.py", line 107, in async_turn_on
    await self._execute_command(method="SET",
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/device.py", line 281, in _execute_command
    return await self._manager.async_execute_cmd(destination_device_uuid=self.uuid,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 838, in async_execute_cmd
    return await self.async_execute_cmd_client(client=client,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 903, in async_execute_cmd_client
    response = await self._async_send_and_wait_ack(
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 934, in _async_send_and_wait_ack
    raise CommandTimeoutError(message=str(message), target_device_uuid=target_device_uuid, timeout=timeout)
meross_iot.model.exception.CommandTimeoutError
```
At this point I restated the integration
```
2021-08-29 09:12:18 WARNING (MainThread) [custom_components.meross_cloud] Skip MQTT cert validation option set to: True
2021-08-29 09:12:19 WARNING (MainThread) [custom_components.meross_cloud] 
===============================
Meross Cloud Custom component
Developed by Alberto Geniola
Low level library version: 0.4.2.2
-------------------------------
This custom component is under development and not yet ready for production use.
In case of errors/misbehave, please report it here: 
https://github.com/albertogeniola/meross-homeassistant/issues

If you like this extension and you want to support it, please consider donating.
-------------------------------
List of devices reported by HTTP API:
- Applique Porte Entrée Droite (msl120d) - OnlineStatus.ONLINE
- Applique Porte Entrée Gauche (msl120d) - OnlineStatus.ONLINE
- Lampe Chevet Tony (msl430) - OnlineStatus.ONLINE
- Potelet Allée (msl120d) - OnlineStatus.ONLINE
- Potelet Terrasse (msl120d) - OnlineStatus.ONLINE
- Portail (msg100) - OnlineStatus.ONLINE
- Routeur Internet (mss310) - OnlineStatus.ONLINE
- Home Cinéma (mss310) - OnlineStatus.ONLINE
- Applique Office (mss710) - OnlineStatus.ONLINE
- Prise Salle à Manger (mss310) - OnlineStatus.ONLINE
- Plafonnier Salle à Manger (mss550l) - OnlineStatus.ONLINE
- Lampe Chevet Stéphane (msl420) - OnlineStatus.ONLINE
- Ordinateur Tony (mss310) - OnlineStatus.ONLINE
- Ordinateur Stephane (mss310) - OnlineStatus.ONLINE
===============================
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Portail (#BASE:2008140285600236100348e1e92b246a) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Prise Salle à Manger (#BASE:18122136570906251h0134298f1450de) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Potelet Terrasse (#BASE:2010234176931990831948e1e9379e65) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Potelet Allée (#BASE:2010234169702290831948e1e937b681) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Routeur Internet (#BASE:18071993274327251a0534298f1371ee) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Ordinateur Tony (#BASE:18071935192641251a0534298f136e8d) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Lampe Chevet Stéphane (#BASE:1812141808483029088034298f19559e) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Plafonnier Salle à Manger (#BASE:1906134950568839084648e1e9501f4b) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Lampe Chevet Tony (#BASE:2101262970091890841048e1e94908bf) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Device Ordinateur Stephane (#BASE:18071948351637251a0534298f13736d) has been already added to the registry.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Applique Porte Entrée Droite (msl120d, HW 2.0.0, FW 2.1.2, class: msl120d:2.0.0:2.1.2) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Applique Office (mss710, HW 3.0.0, FW 3.1.6, class: mss710:3.0.0:3.1.6) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Applique Porte Entrée Gauche (msl120d, HW 2.0.0, FW 2.1.2, class: msl120d:2.0.0:2.1.2) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Home Cinéma (mss310, HW 1.0.0, FW 1.1.18, class: mss310:1.0.0:1.1.18) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Prise Salle à Manger (mss310, HW 2.0.0, FW 2.1.12, class: mss310:2.0.0:2.1.12) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Ordinateur Tony (mss310, HW 1.0.0, FW 1.1.18, class: mss310:1.0.0:1.1.18) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Routeur Internet (mss310, HW 1.0.0, FW 1.1.18, class: mss310:1.0.0:1.1.18) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Plafonnier Salle à Manger (mss550l, HW 2.0.0, FW 2.1.6, class: mss550l:2.0.0:2.1.6) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Potelet Allée (msl120d, HW 2.0.0, FW 2.1.2, class: msl120d:2.0.0:2.1.2) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Potelet Terrasse (msl120d, HW 2.0.0, FW 2.1.2, class: msl120d:2.0.0:2.1.2) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Portail (msg100, HW 3.5.0, FW 3.2.3, class: msg100:3.5.0:3.2.3) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Lampe Chevet Stéphane (msl420, HW 2.0.0, FW 2.1.11, class: msl420:2.0.0:2.1.11) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Lampe Chevet Tony (msl430, HW 4.0.0, FW 4.1.13, class: msl430:4.0.0:4.1.13) that has become online while we were offline.
2021-08-29 09:12:20 WARNING (MainThread) [meross_iot.manager] Found a new device Ordinateur Stephane (mss310, HW 1.0.0, FW 1.1.18, class: mss310:1.0.0:1.1.18) that has become online while we were offline.
2021-08-29 09:12:25 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.prise_salle_a_manger_mss310_power_sensor_a_0 is taking over 10 seconds
2021-08-29 09:12:25 ERROR (MainThread) [meross_iot.manager] Timeout occurred while waiting a response for message b'{"header": {"from": "/app/55147-dce9f6b40e9066af8782277ad98dd10d/subscribe", "messageId": "8d4ce63a253adc1c9647e780ba255723", "method": "GET", "namespace": "Appliance.Control.Electricity", "payloadVersion": 1, "sign": "47c3207c03b8d8a8af8d516484eb91da", "timestamp": 1630221136}, "payload": {"channel": 0}}' sent to device uuid 18122136570906251h0134298f1450de. Timeout was: 10.0 seconds. Mqtt Host: mqtt-eu.meross.com:443.Global manager stats (last minute): Issued -> 5, Delayed -> 0, Dropped -> 0
2021-08-29 09:12:25 ERROR (MainThread) [custom_components.meross_cloud.sensor] Error occurred.
-------------------------------------
Component version: 0.4.2.2
Device info: 
	Name: Prise Salle à Manger
	UUID: 18122136570906251h0134298f1450de
	Type: mss310
	HW Version: 2.0.0
	FW Version: 2.1.12
Error Message: "An exception occurred"
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 919, in _async_send_and_wait_ack
    return await asyncio.wait_for(future, timeout, loop=self._loop)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/meross_cloud/sensor.py", line 365, in async_update
    await self._device.async_get_instant_metrics(channel=self._channel_id)
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/mixins/electricity.py", line 43, in async_get_instant_metrics
    result = await self._execute_command(method="GET",
  File "/usr/local/lib/python3.9/site-packages/meross_iot/controller/device.py", line 281, in _execute_command
    return await self._manager.async_execute_cmd(destination_device_uuid=self.uuid,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 838, in async_execute_cmd
    return await self.async_execute_cmd_client(client=client,
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 903, in async_execute_cmd_client
    response = await self._async_send_and_wait_ack(
  File "/usr/local/lib/python3.9/site-packages/meross_iot/manager.py", line 934, in _async_send_and_wait_ack
    raise CommandTimeoutError(message=str(message), target_device_uuid=target_device_uuid, timeout=timeout)
meross_iot.model.exception.CommandTimeoutError
```
So there is a few asyncio.exceptions.CancelledError / TimeoutError and then it resume operation normally until the next hiccup.

I would gadly help you resolve this issue event though I'm not a Python expert. From what I can read it seems other users are having similar issues (so feel free to attach this issue to the most appropriate) 

Note : I managed once to lock my meross account with the Domoticz plugin because this plugin was requesting token each time it starts so while I was testing some modifications I made to this plugin I ran out of allowed token. Other than that I never experience any other locking out.

Best Regards,

**Comment by StephaneM60** [comment]
FYI this morning the integration was not working with the same errors. But this time no Internet event occured during the night (it is possible though that some meross devices changed access point during the night : when the blinds close many devices are switching from an access point to another one)

Anyway : HA was not in a good sate, 56% CPU, as soon as I disable the meross integration HA CPU usage returned to normal.

At first I suspected ffmepg for the high CPU usage, so I disabled it yesterday, so it is the meross integration that does not behave correctly.

**Comment by StephaneM60** [comment]
I managed to make this integration much more stable. I compared your code with the code of the domoticz Integration I was using and noticed that you use a 10 sec. refresh rate (for power devices) and the domoticz was using 1 min. refresh rate.

So I modified common.py HA_SENSOR_POLL_INTERVAL_SECONDS to be 60 and SENSOR_SAMPLE_CACHE_INTERVAL_SECONDS to be 120.

I modified as well sensor.py to use HA_SENSOR_POLL_INTERVAL_SECONDS in place of the hardcoded 10 sec delay that is used in some places like in class PowerSensorWrapper etc...

With these modifications : the integration works very well, though it still doesn't survive my router restart.
